### PR TITLE
[dynamic_layouts] Add base classes for building dynamic grids

### PR DESCRIPTION
**Pull request by Piinks**
This adds the base upon which @DavBot02 and @snat-s will be adding their dynamic grid layouts.
This package will remain unpublished while their changes are added.
See their design documents here:

- [stagger grid design doc](https://docs.google.com/document/d/1PCakLu5t1JtZ0cQbhYWw0ZnSPa0s7CUVAgG0dqde1qM/edit?usp=sharing&resourcekey=0-RAGxfw9E6SHfJ6jKnG7NFw)
  - https://github.com/flutter/flutter/issues/40856 
- [wrap grid design doc](https://docs.google.com/document/d/1auZtWT1T5CkUKOUTUld8rF91iN7PvQoOF0IlBNXiDSo/edit?usp=sharing)
  - https://github.com/flutter/flutter/issues/97544

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like some checks are unhappy. You're probably missing a run of the dart formatter?

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+    bool reachedEnd = false;
+
+    // This algorithm in principle is straight-forward: find the first child
+    // that overlaps the given scrollOffset, creating more children at the top
+    // of the grid if necessary, then walk through the grid updating and laying
+    // out each child and adding more at the end if necessary until we have
+    // enough children to cover the entire viewport.
+    //
+    // It is complicated by one minor issue, which is that any time you update
+    // or create a child, it's possible that the some of the children that
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:78]: nit: that the some?

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    // inconsistent state, and requiring that missing nodes be recreated.
+    //
+    // To keep this mess tractable, this algorithm starts from what is currently
+    // the first child, if any, and then walks up and/or down from there, so
+    // that the nodes that might get removed are always at the edges of what has
+    // already been laid out.
+
+    // Make sure we have at least one child to start from.
+    if (firstChild == null) {
+      if (!addInitialChild()) {
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:89]: could be one if?

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        // There are no children.
+        geometry = SliverGeometry.zero;
+        childManager.didFinishLayout();
+        return;
+      }
+    }
+
+    // We have at least one child.
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:97]: assert that firstChild is not null to support that claim? :)

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    if (childScrollOffset(firstChild!) == null) {
+      int leadingChildrenWithoutLayoutOffset = 0;
+      while (earliestUsefulChild != null && childScrollOffset(earliestUsefulChild) == null) {
+        earliestUsefulChild = childAfter(earliestUsefulChild);
+        leadingChildrenWithoutLayoutOffset += 1;
+      }
+      // We should be able to destroy children with null layout offset safely,
+      // because they are likely outside of viewport
+      collectGarbage(leadingChildrenWithoutLayoutOffset, 0);
+      // If can not find a valid layout offset, start from the initial child.
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:120]: The if clause may be missing a subject?

Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:120]: can not -> cannot?

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+      if (firstChild == null) {
+        if (!addInitialChild()) {
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:122]: could be one if?

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  @override
+  double childCrossAxisPosition(RenderBox child) {
+    final SliverGridParentData childParentData = child.parentData! as SliverGridParentData;
+    return childParentData.crossAxisOffset!;
+  }
+
+  @override
+  void performLayout() {
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:56]: Is a lot of this logic shared with e.g. RenderSliverGrid? If true, wondering, if we could find a way to share that better. We don't have to necessarily do it in this PR, though. But maybe add a TODO if you think it is possible...

**Review by goderbauer — COMMENTED** [review]
Not necessarily in this PR, but we should also add an `example` directory with an example app showcasing these grids. That helps developers get started and also gives the package more points on pub.dev! cc @DavBot02 @snat-s

**Comment by Piinks** [comment]
> Not necessarily in this PR, but we should also add an example directory with an example app showcasing these grids.

You read my mind. :)

File: packages/dynamic_layouts/analysis_options.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+include: package:flutter_lints/flutter.yaml
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/dynamic_layouts/analysis_options.yaml:1]: Oh right, thanks

File: packages/dynamic_layouts/lib/src/base_grid_layout.dart [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  /// required size.
+  ///
+  /// If the [mainAxisExtent] is [double.infinity] the child will be allowed to
+  /// chose its own size in the main axis. Similarly, an infinite
+  /// [crossAxisExtent] will result in the child sizing itself in the cross
+  /// axis. Otherwise, the provided cross axis size or the
+  /// [SliverConstraints.crossAxisExtent] will be used to create tight
+  /// constraints in the cross axis.
+  @override
+  BoxConstraints getBoxConstraints(SliverConstraints constraints) {
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/dynamic_layouts/lib/src/base_grid_layout.dart:30]: Ah great idea. The main difference here is that SliverGridGeometry.getBoxConstraints only allows them to be tight constraints.

File: packages/dynamic_layouts/lib/src/stagger_layout.dart [hunk 9]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/dynamic_layouts/lib/src/stagger_layout.dart:1]: Ok! I will remove them. @DavBot02 & @snat-s, these were placeholders for your code, you can refer back to them here when you go to add it.

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart (hunk 2 above)
Comment (Piinks) [reply packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:89]: Sure! I think originally I was following the same pattern in RenderSliverList so it is easy to follow and maintain the similar logic between them, maybe I can do a clean up in RenderSliverList later.

File: packages/dynamic_layouts/lib/src/base_grid_layout.dart [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+          minHeight: crossMinExtent,
+          maxHeight: crossAxisExtent,
+          minWidth: mainMinExtent,
+          maxWidth: mainAxisExtent,
+        );
+    }
+  }
+}
+
+/// Manages the size and position of all the tiles in a [RenderSliverGrid].
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/dynamic_layouts/lib/src/base_grid_layout.dart:54]: Added a TODO for follow up PRs.

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart (hunk 6 above)
Comment (Piinks) [reply packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:56]: performLayout is really where it differs from RenderDynamicSliverGrid, I'll clean it up a bit.

File: packages/dynamic_layouts/lib/src/base_grid_layout.dart [hunk 11]
```diff
@@ -16,23 +16,27 @@ class DynamicSliverGridGeometry extends SliverGridGeometry {
...
   ///
   /// If the [mainAxisExtent] is [double.infinity] the child will be allowed to
-  /// chose its own size in the main axis. Similarly, an infinite
+  /// choose its own size in the main axis. Similarly, an infinite
   /// [crossAxisExtent] will result in the child sizing itself in the cross
   /// axis. Otherwise, the provided cross axis size or the
   /// [SliverConstraints.crossAxisExtent] will be used to create tight
   /// constraints in the cross axis.
+  ///
+  /// This differs from [SliverGridGeometry.getBoxConstraints] in that is allows
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/base_grid_layout.dart:30]: is -> it

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 12]
```diff
@@ -51,21 +52,21 @@ void main() {
...
-            }
-          ),
+              gridDelegate: TestDelegate(crossAxisCount: 2),
+              itemCount: 3,
+              itemBuilder: (BuildContext context, int index) {
+                return SizedBox.square(
+                  dimension: TestSimpleLayout.childExtent,
+                  child: Text('Index $index'),
+                );
+              }),
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:69]: add a trailing comma for nicer formatting

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 13]
```diff
@@ -87,14 +88,13 @@ void main() {
...
-              );
-            }
-          ),
+              gridDelegate: TestDelegate(crossAxisCount: 2),
+              itemBuilder: (BuildContext context, int index) {
+                return SizedBox.square(
+                  dimension: TestSimpleLayout.childExtent,
+                  child: Text('Index $index'),
+                );
+              }),
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:97]: same

**Review by goderbauer — COMMENTED** [review]
not a full review, just some quick nits found between meetings.

File: packages/dynamic_layouts/.gitignore [hunk 14]
```diff
@@ -0,0 +1,30 @@
+# Miscellaneous
```
Comment (goderbauer) [inline packages/dynamic_layouts/.gitignore:1]: Can we just use the global `.gitignore` [1] or do we need a special one?

[1] https://github.com/flutter/packages/blob/main/.gitignore

File: packages/dynamic_layouts/lib/src/dynamic_grid.dart [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  @override
+  RenderDynamicSliverGrid createRenderObject(BuildContext context) {
+    final SliverMultiBoxAdaptorElement element =
+        context as SliverMultiBoxAdaptorElement;
+    return RenderDynamicSliverGrid(
+        childManager: element, gridDelegate: gridDelegate);
+  }
+
+  @override
+  void updateRenderObject(
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/dynamic_grid.dart:71]: nit: dart-format formats these nicer if you add a trailing comma to the parameter list.

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+      MaterialApp(
+        home: Scaffold(
+          body: DynamicGridView(
+            gridDelegate: TestDelegate(crossAxisCount: 2),
+            children: const <Widget>[],
+          ),
+        ),
+      ),
+    );
+    await tester.pump();
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:24]: Why the extra pump?

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+              (int index) => SizedBox.square(
+                dimension: TestSimpleLayout.childExtent,
+                child: Text('Index $index'),
+              ),
+            ),
+          ),
+        ),
+      ),
+    );
+    await tester.pump();
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:42]: same.

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+              return SizedBox.square(
+                dimension: TestSimpleLayout.childExtent,
+                child: Text('Index $index'),
+              );
+            },
+          ),
+        ),
+      ),
+    );
+    await tester.pump();
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:74]: same

File: packages/dynamic_layouts/test/dynamic_grid_test.dart [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
```
Comment (goderbauer) [inline packages/dynamic_layouts/test/dynamic_grid_test.dart:103]: same.

File: packages/dynamic_layouts/lib/src/render_dynamic_grid.dart [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+    //     tile to refer back to, but that would mean tiles could not change in
+    //     size or be added/removed without having to recompute the whole layout.
+    //
+    // 2 - Currently, Flutter only supports collecting garbage in sequential
+    //     order. Dynamic layout patterns can break this assumption. In order to
+    //     truly collect garbage efficiently, support for nonsequential garbage
+    //     collection is necessary.
+    // TODO(all): document this behavior so users are aware of this temporary
+    //   also - file a bug in https://github.com/flutter/flutter/labels/p%3A%20dynamic_layouts
+    // trade off
```
Comment (goderbauer) [inline packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:355]: The "trade off" appears lonely? Should this be here?

Comment (goderbauer) [reply packages/dynamic_layouts/lib/src/render_dynamic_grid.dart:355]: Oh, I think this belong to the end of line 353. Basically, 354 needs to move down one line.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/dynamic_layouts/.gitignore (hunk 14 above)
Comment (Piinks) [reply packages/dynamic_layouts/.gitignore:1]: Oh I was looking at adaptive_scaffold for comparison, it has the same `.gitignore`, should they be removed?

**Pull request merged**